src/python/pants/backend/build_files/fmt/buildifier/rules_integration_test.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.build_files.fmt.buildifier.rules import BuildifierRequest
+from pants.backend.build_files.fmt.buildifier.rules import rules as buildifier_rules
+from pants.backend.codegen.protobuf.target_types import rules as target_types_rules
+from pants.core.goals.fmt import FmtResult
+from pants.core.util_rules import external_tool
+from pants.engine.fs import CreateDigest, FileContent, PathGlobs
+from pants.engine.internals.native_engine import Digest, Snapshot
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+class Materials:
+    def __init__(self, **kwargs):
+        pass
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        rules=[
+            *buildifier_rules(),
+            *external_tool.rules(),
+            *target_types_rules(),
+            QueryRule(FmtResult, [BuildifierRequest]),
+        ],
+        # NB: Objects are easier to test with
+        objects={"materials": Materials},
+    )
+
+
+GOOD_FILE = dedent(
+    """\
+    materials(
+        drywall = 40,
+        status = "paid",
+        studs = 200,
+    )
+    """
+)
+
+BAD_FILE = dedent(
+    """\
+    materials(status='paid', studs=200, drywall=40)
+    """
+)
+
+
+def run_buildifier(rule_runner: RuleRunner) -> FmtResult:
+    rule_runner.set_options(
+        ["--backend-packages=pants.backend.build_files.fmt.buildifier"],
+        env_inherit={"PATH", "PYENV_ROOT"},
+    )
+    snapshot = rule_runner.request(Snapshot, [PathGlobs(["**/BUILD"])])
+    fmt_result = rule_runner.request(FmtResult, [BuildifierRequest(snapshot)])
+    return fmt_result
+
+
+def get_snapshot(rule_runner: RuleRunner, source_files: dict[str, str]) -> Snapshot:
+    files = [FileContent(path, content.encode()) for path, content in source_files.items()]
+    digest = rule_runner.request(Digest, [CreateDigest(files)])
+    return rule_runner.request(Snapshot, [digest])
```

Comment:
Can use https://github.com/pantsbuild/pants/blob/d0c52104c0b963119fdc5320df2c64f7df7abfcd/src/python/pants/testutil/rule_runner.py#L454-L465 for this.